src/Build/BackEnd/Components/RequestBuilder/Lookup.cs

```diff
@@ -1072,10 +1072,13 @@ public void ApplyModifications(MetadataModifications other)
                 if (other._keepOnlySpecified)
                 {
                     // Any metadata not specified in other must be removed from this one.
-                    var metadataToRemove = new List<string>(_modifications.Keys.Where(m => other[m].Remove));
+                    var metadataToRemove = new List<string>(_modifications.Keys);
                     foreach (var metadata in metadataToRemove)
                     {
-                        _modifications.Remove(metadata);
+                        if (other[metadata].Remove)
+                        {
+                            _modifications.Remove(metadata);
+                        }
```

Comment:
I felt like this could be optimized, and this went through a few iterations:
First was an attempt to remove Linq, since it allocates a little extra, and just Where the dictionary, but then I needed to allocate a new dictionary with ToDictionary.
I also tried inlining the filter without creating an array, but that failed because it's illegal to modify the dictionary while enumerating over it.
Then I fixed that last part by making the list but didn't change that I was checking Remove later.

I think you're ultimately right, and none of those changes were actually helpful. I'll revert this part. Thanks!